crypto/ascon/ascon_aead128.c

```diff
@@ -0,0 +1,436 @@
+        q1 = x3 & (~x2);            \
+        x1 ^= q1;                   \
+        x3 ^= q0;                   \
+        x1 ^= x0;                   \
+        x3 ^= x2;                   \
+        x0 ^= x4;                   \
+        x2 = ~x2;                   \
+    } while (0)
+
+/* linear layer, NIST SP 800-232 Figure 3 */
+#define ASCONPL(x0, x1, x2, x3, x4)          \
+    do {                                     \
+        x0 ^= ROR64(x0, 19) ^ ROR64(x0, 28); \
+        x1 ^= ROR64(x1, 61) ^ ROR64(x1, 39); \
+        x2 ^= ROR64(x2, 1) ^ ROR64(x2, 6);   \
+        x3 ^= ROR64(x3, 10) ^ ROR64(x3, 17); \
+        x4 ^= ROR64(x4, 7) ^ ROR64(x4, 41);  \
+    } while (0)
+
+/* one round */
+#define ASCONP1(x0, x1, x2, x3, x4, rcon)  \
+    do {                                   \
+        ASCONPC(x0, x1, x2, x3, x4, rcon); \
+        ASCONPS(x0, x1, x2, x3, x4);       \
+        ASCONPL(x0, x1, x2, x3, x4);       \
+    } while (0)
+
+/* 8 rounds */
+#define ASCONP8(x0, x1, x2, x3, x4)           \
+    do {                                      \
+        ASCONP1(x0, x1, x2, x3, x4, 0xB4ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0xA5ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x96ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x87ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x78ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x69ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x5AULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x4BULL); \
+    } while (0)
+
+/* 12 rounds */
+#define ASCONP12(x0, x1, x2, x3, x4)          \
+    do {                                      \
+        ASCONP1(x0, x1, x2, x3, x4, 0xF0ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0xE1ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0xD2ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0xC3ULL); \
+        ASCONP8(x0, x1, x2, x3, x4);          \
+    } while (0)
+
+/* misc ascon flags for the context */
+#define ASCONFLG_AAD 0x0000000000000001ULL /* has AAD inputs? */
+#define ASCONFLG_DEC 0x0000000000000002ULL /* in decrypt mode? */
+#define ASCONFLG_DOMAINSEP 0x8000000000000000ULL /* ready to absorb non-AAD? */
+
+/* omit test harness if this is openssl-internal */
+#ifndef OPENSSL_BUILDING_OPENSSL
+# include <stdio.h>
+
+void ascon_test_state()
+{
+    uint64_t s0 = 0, s1 = 0, s2 = 0, s3 = 0, s4 = 0;
+
+    s4 = 0x0000080100CC0002ULL;
+    ASCONP12(s0, s1, s2, s3, s4);
+    printf("%lx%lx%lx%lx%lx\n", s0, s1, s2, s3, s4);
+}
+
+#endif
+
+static ASCON_INLINE void ascon_aead128_update(ascon_aead128_ctx *ctx,
+                                              unsigned char *out,
+                                              const unsigned char *in,
+                                              size_t len)
+{
+    uint64_t s0, s1, s2, s3, s4, flags;
+    unsigned char pad = 0x01;
+
+    if (ctx->flags & ASCONFLG_DOMAINSEP) {
+        flags = ctx->flags;
+        if (flags & ASCONFLG_AAD) {
+            ctx->flags = 0;
+            ascon_aead128_update(ctx, NULL, &pad, 1);
```

Comment:
I am really not a fan of mutating the state and doing a recursive call - There should be a nice way of doing this without recursion.